### PR TITLE
feat: implement plugin with helper function

### PR DESCRIPTION
**Pull request by anshbajpai01**
## Purpose

Defines the plugin as a ["library plugin"][library-plugins] that adds a `chinmina_token` utility script to the path. The utility can be referenced in steps in order to get a GitHub token for use in calls to the GitHub API.

For example, this can be used to get release assets from GitHub using the `gh` CLI:

```shell
GITHUB_TOKEN="$(chinmina-token "profile-name")" gh releases download ...
```
[library-plugins]: https://buildkite.com/docs/pipelines/integrations/plugins/writing#designing-plugins-single-command-plugins-versus-library-plugins

## Context

- [CSRE-5541](https://cultureamp.atlassian.net/jira/software/c/projects/CSRE/boards/412?assignee=712020%3A1107d822-435c-4fe3-8924-c5384e7a0251&selectedIssue=CSRE-5541)

## Summary by CodeRabbit

- **New Features**
  - Introduced a Buildkite plugin to fetch Chinmina tokens for organization profiles using cached OIDC tokens.
  - Added a command-line tool to obtain authentication tokens for specified profiles.
  - Enabled environment-specific configuration through new environment files.
  - Extended environment path setup for improved script execution in hooks.

- **Documentation**
  - Added a comprehensive README with setup, configuration, usage examples, and contribution guidelines.

- **Tests**
  - Implemented automated tests to verify token fetching, environment setup, and error handling.

- **Chores**
  - Added Docker Compose setup for local testing.
  - Updated gitignore to exclude sensitive configuration files.

### CODE REVIEW

File: bin/chinmina_token [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+# This script is responsible for getting a token for a given profile (or the default)
+profile_name=$1
+oidc_cached_path="/tmp/oidc_auth_token_${BUILDKITE_JOB_ID}.cache"
```
Comment (liamstevens) [inline bin/chinmina_token:5]: It is going to be worth keeping this in mind in case delay between checkout and this is large enough to cause issues (to my mind, this is unlikely).

Good piggybacking off the existing agent behaviour though.

File: bin/chinmina_token [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+if [[ ${profile_name} == "default" || ${profile_name} == "repo:default" ]]; then
+  path="git-credentials"
+else
+  path="organization/git-credentials/${profile_name}"
+fi
+
```
Comment (liamstevens) [inline bin/chinmina_token:12]: This isn't quite right. 

Git-credentials will return a JSON-formatted set of information that **includes** the token.

For naked token requests, it's better to hit the `/token` endpoint. See t[his line in most recent merged PR to our milestone branch](https://github.com/chinmina/chinmina-bridge/pull/80/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R67).

The structure of the token including the JSON keys that it is serialised to can be found [here](https://github.com/chinmina/chinmina-bridge/pull/80/files#diff-dd24b0f1dbf4030c56eba23df72e85b0076400382ec6b8541d92f05f6ac25bffR27).

You can see how the `git-credentials` plugin handles them [here](https://github.com/chinmina/chinmina-bridge/pull/80/files#diff-56307bdb7ded67ebe516f562b90384407b21b2d34e8b5766b42602e1fb9705f0R94).

We aren't tied to this format, I'm open to us making changes to the returned token from this endpoint.

File: plugin.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+requirements: []
+configuration:
+  properties:
+    chinmina-url:
+      type: string
+      description: The URL of the Chinmina Bridge agent that creates a token for a pipeline.
+    audience:
+      type: string
+      description: |
+        (Default `chinmina:default`.) The audience to use for the Buildkite OIDC
```
Comment (liamstevens) [inline plugin.yml:18]:
```suggestion
        (Default `chinmina:default`) The audience to use for the Buildkite OIDC
```

File: plugin.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        JWT that is sent to the vendor agent. Must match the setting in the
+        vendor agent.
+    profiles:
+      type: array
```
Comment (liamstevens) [inline plugin.yml:22]: Is this type right? It doesn't look like we are using arrays elsewhere.

File: tests/download.bats [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+# Uncomment the following line to debug stub failures
+# export [stub_command]_STUB_DEBUG=/dev/tty
+#export DOCKER_STUB_DEBUG=/dev/tty
+
+setup(){
+
+    TEST_ROOT=$(mktemp -d)
+
+    #Path to the shim
+    SHIM_DIR="$BATS_TEST_DIRNAME/../lib"
```
Comment (liamstevens) [inline tests/download.bats:16]: References to shim should be removed with this new approach, no?

File: tests/download.bats [hunk 6]
```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env bats
```
Comment (liamstevens) [inline tests/download.bats:1]: This file looks like it's out of date and references the previous approach.

File: bin/chinmina_token (hunk 1 above)
Comment (n-tucker) [reply bin/chinmina_token:6]: This is a good callout, since getting the BK OIDC token happens in the agent environment hook I'd expect this to occur before this script is executed, but it's worth double checking in a smoke test

File: bin/chinmina_token [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+# This script is responsible for getting a token for a given profile (or the default)
+profile_name=$1
+oidc_cached_path="/tmp/oidc_auth_token_${BUILDKITE_JOB_ID}.cache"
+url="https://chinmina.usw2.ci.cultureamp.io" # TODO: Make 'url' configurable from plugin inputs.
+args="" # No data is required to fetch the token
+
+if [[ ${profile_name} == "default" || ${profile_name} == "repo:default" ]]; then
+  path="token"
+else
+  path="organization/token/${profile_name}"s
```
Comment (n-tucker) [inline bin/chinmina_token:13]: I assume this is a typo?

File: bin/chinmina_token (hunk 1 above)
Comment (liamstevens) [reply bin/chinmina_token:6]: As discussed in the huddle we do want to rework this somewhat as this introduces a requirement for consumers of this (open source) plugin to configure their agents in a way similar to what we have done at CA.

File: bin/chinmina_token (hunk 2 above)
Comment (anshbajpai01) [reply bin/chinmina_token]: Resolved now. Thank you !

File: plugin.yml (hunk 4 above)
Comment (anshbajpai01) [reply plugin.yml]: Updated.

File: bin/chinmina_token [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  --header "Authorization: Bearer ${oidc_auth_token}" \
+  --header "Content-Type: text/plain" \
+  --header "Accept: text/plain")
+
+echo "$chinmina_response" | jq .
+token_exists=$(echo "$chinmina_response" | jq 'has("token")')  
+
+if $token_exists; then
+    chinmina_token=$(echo "$chinmina_response" | jq '.token')
+    if [ ${#chinmina_token} -eq 0 ]; then
```
Comment (liamstevens) [inline bin/chinmina_token:32]: typo?

Comment (anshbajpai01) [reply bin/chinmina_token:32]: Are you referring to the if condition ?

Comment (liamstevens) [reply bin/chinmina_token:32]: Ah, just re-read the condition that is being checked. All good.

**Comment by liamstevens** [comment]
Looking pretty good. Needs some tidying up (both in commit history and in some older content), and the open question around the use of a cached OIDC token to be resolved.

File: bin/chinmina_token (hunk 1 above)
Comment (n-tucker) [reply bin/chinmina_token:5]: We could simply copy/paste [existing logic from the credential helper plugin](https://github.com/chinmina/chinmina-git-credentials-buildkite-plugin/blob/main/credential-helper/buildkite-connector-credential-helper#L16-L27) to avoid assumptions around agent setup

Comment (liamstevens) [reply bin/chinmina_token:5]: +1, though I will leave the decision up to @anshbajpai01

Comment (liamstevens) [reply bin/chinmina_token:5]: As discussed in huddle this should be a copy-paste from the `git-credentials` script.

File: README.md [hunk 9]
```diff
@@ -0,0 +1,6 @@
+<!-- ./command.sh curl https://github.com/cultureamp/deploy-buildkite-plugin/releases/download/v0.1/deploy-buildkite-plugin_darwin_arm64 -->
+
+# TODO
+
+- BATS tests
+- Add shimming capabilities so that script can transparently replace curl and wget
```
Comment (liamstevens) [inline README.md:6]: Let's tidy this up.

File: bin/chinmina_token (hunk 8 above)
Comment (liamstevens) [reply bin/chinmina_token:45]: @n-tucker that would be the case if we were hitting the `git-credentials` endpoint. We're not doing that, and the `/token` endpoint should handle those sorts of cases slightly differently to account for this sort of edge case. 

Looking at the code there, it looks like this is not **not** the case and it will need to be updated. I've created an issue for it and will bring it into our release branch [see here](https://github.com/chinmina/chinmina-bridge/issues/81)

Comment (liamstevens) [reply bin/chinmina_token:45]: Skim-reading strikes again.

Yes, this implementation handles the empty responses, so no change is needed here for now.

I do think that we should respond differently on the `/token` endpoint though so I will leave the issue up. There will need to be a little bit of rework here should we modify that behaviour.

File: bin/chinmina_token [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  --header "Authorization: Bearer ${oidc_auth_token}" \
+  --header "Content-Type: text/plain" \
+  --header "Accept: text/plain")
+
+token_exists=$(echo "$chinmina_response" | jq 'has("token")')  
+
+if $token_exists; then
+    chinmina_token=$(echo "$chinmina_response" | jq '.token')
+    if [ ${#chinmina_token} -eq 0 ]; then
+        echo "request failed: token doesn't exist"
```
Comment (jamestelfer) [inline bin/chinmina_token:42]: All output apart from the token needs to be on stderr

File: bin/chinmina_token [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        exit 1
+    fi
+else
+    echo "request failed: token not found"
+    exit 1
+fi 
+
+echo $chinmina_token | tr -d '"'
```
Comment (jamestelfer) [inline bin/chinmina_token:50]: Use `jq -r '.token'` above to get raw output without quotes instead.

File: bin/chinmina_token [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+chinmina_response=$(curl --silent --show-error --fail \
+  --request POST "${url}/${path}" \
+  --data "${args}" \
+  --header "Authorization: Bearer ${oidc_auth_token}" \
+  --header "Content-Type: text/plain" \
+  --header "Accept: text/plain")
+
+token_exists=$(echo "$chinmina_response" | jq 'has("token")')  
+
+if $token_exists; then
```
Comment (jamestelfer) [inline bin/chinmina_token:39]: This command outputs the literal string `true` or `false`. Let's change the condition to be 

```suggestion
token_exists="$(echo "$chinmina_response" | jq 'has("token")')"

if [[ "${token_exists}" == "true" ]]; then
```

... and then we can invert the test here to have less nesting:

```suggestion
token_exists="$(echo "$chinmina_response" | jq 'has("token")')"

if [[ "${token_exists}" == "false" ]]; then
  1>&2 echo "request failed: no token returned in Chinmina response"
```

File: bin/chinmina_token (hunk 11 above)
Comment (jamestelfer) [inline bin/chinmina_token:50]:
```suggestion
echo "${chinmina_token}"
```

File: bin/chinmina_token [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    chinmina_token=$(echo "$chinmina_response" | jq '.token')
```
Comment (jamestelfer) [inline bin/chinmina_token:40]:
```suggestion
    chinmina_token=$(echo "$chinmina_response" | jq -r '.token')
```

This needs to use "raw" output so that quotes aren't included around the string. This removes the need for `tr` and means that the string length test will fail if the token field is empty.

File: bin/chinmina_token [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    if [ ${#chinmina_token} -eq 0 ]; then
```
Comment (jamestelfer) [inline bin/chinmina_token:41]:
```suggestion
    if [[ ${#chinmina_token} -eq 0 ]]; then
```
nit: consistency

File: bin/chinmina_token (hunk 10 above)
Comment (jamestelfer) [inline bin/chinmina_token:42]:
```suggestion
        1>&2 echo "request failed: token doesn't exist"
```

This might seem like a nit, but this script is designed to have its output captured to a variable. It's important to be careful with output to `stdout` in scripts.

**Review by jamestelfer — CHANGES_REQUESTED** [review]
Refinement changes, overall functionality is clear 👍🏻

File: tests/chinmina_token.bats [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+    # TEST_ROOT=$(mktemp -d)
+
+    #Path to the cache file
+    CACHE_FILE="/tmp/oidc_auth_token_${BUILDKITE_JOB_ID}.cache"
+    echo "Name is: $CACHE_FILE"
+    echo "$OIDC_TOKEN" > "$CACHE_FILE"
+    export CACHE_FILE
+
+}
```
[the remark this answers is not in the document]
Comment (anshbajpai01) [reply tests/chinmina_token.bats:23]: It doesn't need to be created every time, as of now it gets created at the start of test and deleted during teardown. Should I just create a persistent file instead in the tests directory?

File: README.md [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    command: |
+    	export GH_TOKEN=$$(chinmina_token "org:profile-name")
+      gh release download ${tag} -R ${org}/${repo} --pattern "{filename}.zip"
+```
+
+To use the token and fetch private github releases, usage would be the following:
+
+```bash
+
+```
```
Comment (n-tucker) [inline README.md:24]: Is there meant to be an example here? The above example looks sufficient so I think this can be removed

File: README.md [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+```yml
+steps:
+  - plugins:
+      - chinmina/chinmina-tokem#v1.0.0:
+          chinmina-url: "https://chinmina-bridge-url"
+          audience: "chinmina:your-github-organization"
+    command: |
+    	export GH_TOKEN=$$(chinmina_token "org:profile-name")
+      gh release download ${tag} -R ${org}/${repo} --pattern "{filename}.zip"
+```
```
Comment (n-tucker) [inline README.md]: Maybe the usage steps can be made clearer? These steps are useful for our own verification, but IMO this should be two separate steps as this is how consumers will use the library plugin:

- Add the following to your `pipeline.yml`, to expose the `chinmina_token` script
```yml
steps:
  - plugins:
      - chinmina/chinmina-tokem#v1.0.0:
          chinmina-url: "https://chinmina-bridge-url"
          audience: "chinmina:your-github-organization"
```

- Add the following to your plugin's hook, to obtain a GitHub token:
```sh
export GH_TOKEN=$$(chinmina_token "org:profile-name")
```

WDYT?

File: README.md [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+token for a pipeline. This is a separate HTTP service that must accessible to
+your Buildkite agents.
+
+### `audience` (string)
+
+**Default:** `chinmina:default`
+
+The value of the `aud` claim of the OIDC JWT that will be sent to
+. This must correlate with the value
+configured in the `chinmina-bridge` settings.
```
[the remark this answers is not in the document]
Comment (therealvio) [reply README.md:40]: Suggestion has the right idea, though specifics may be a little off. More accurately it should be "to the Chinmina service", or something along those lines.

File: bin/chinmina_token [hunk 19]
```diff
@@ -34,17 +34,19 @@ chinmina_response=$(curl --silent --show-error --fail \
...
-else
-    echo "request failed: token not found"
+if [[ "${token_exists}" == "false" ]]; then
+    echo "request failed: no token returned in Chinmina response" >&2
     exit 1
 fi 
 
-echo $chinmina_token | tr -d '"' 
\ No newline at end of file
+if [[ "${token_exists}" == "true" ]]; then
```
Comment (jamestelfer) [inline bin/chinmina_token:44]: This `if` is not required

File: plugin.yml [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+  properties:
+    chinmina-url:
+      type: string
+      description: The URL of the Chinmina Bridge agent that creates a token for a pipeline.
+    audience:
+      type: string
+      description: |
+        (Default `chinmina:default`) The audience to use for the Buildkite OIDC
+        JWT that is sent to the vendor agent. Must match the setting in the
+        vendor agent.
```
[the remark this answers is not in the document]
Comment (therealvio) [reply plugin.yml:19]: Could be useful. "chinmina:default" comes from the service that vends the cached token in the first place, not necessarily for this script. We set audience through a variable reference: `audience="$BUILDKITE_PLUGIN_CHINMINA_TOKEN_LIBRARY_AUDIENCE`.

So in the schema, we can mention this, but I think we don't need to actually set the default value in code, we should leave that responsibility to chinmina, and not inadvertently cause an override in case something changes in a major release for the Chinmina service.

File: tests/chinmina_token.bats [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+    stub curl "echo '{\"profile\": \"profile-name\", \"organisationSlug\": \"org123\", \"token\": \"728282727\", \"expiry\": $(date +%s)}'"
+
+    run './bin/chinmina_token' $profile
+
+    assert_success
+    assert_output --partial "728282727"
+
+    unstub buildkite-agent
+
+}
```
[the remark this answers is not in the document]
Comment (therealvio) [reply tests/chinmina_token.bats]: Not applicable in this case, this test specifically focuses on *not* using a pre-cached OIDC token.

**Review by therealvio — COMMENTED** [review]
Some minor things, otherwise it's looking nice.

File: README.md [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+
+Add the following to your `pipeline.yml`:
+
+```yml
+steps:
+  - plugins:
+      - chinmina/chinmina-token#v1.0.0:
+          chinmina-url: "https://chinmina-bridge-url"
+          audience: "chinmina:your-github-organization"
+```
```
[the remark this answers is not in the document]
Comment (jamestelfer) [reply README.md:26]: There is a Note section above directing folks to the doco, and right below this is an example of how to use this script in a step command.

File: tests/environment.bats [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+
+  local expected_hooks_dir="$(cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd)"
+
+  run_environment "./hooks/environment"
+
+  assert_success
+
+  assert_line 'CHINMINA_TOKEN_LIBRARY_FUNCTION_CHINMINA_URL=test-chinmina-url'
+  assert_line 'CHINMINA_TOKEN_LIBRARY_FUNCTION_AUDIENCE=test-audience'
+  assert_line "PATH=${PATH}:${expected_hooks_dir}"
```
[the remark this answers is not in the document]
Comment (jamestelfer) [reply tests/environment.bats]: Fixed

**Comment by jamestelfer** [comment]
@coderabbitai full review

**Comment by jamestelfer** [comment]
@coderabbitai full review

File: README.md [hunk 24]
```diff
@@ -0,0 +1,85 @@
+# chinmina-token-buildkite-plugin
+
+Adds a `chinimina-token` script to the `PATH`, allowing agent scripts to retrieve
+a GitHub token from Chinmina for the current repository or for an 
+[organizational profile][organization-profiles].
```
[the remark this answers is not in the document]
Comment (liamstevens) [reply README.md:5]: @anshbajpai01 this should be done

File: tests/chinmina_token.bats [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+
+  # values configured by environment hook
+  export CHINMINA_TOKEN_LIBRARY_FUNCTION_CHINMINA_URL="http://sample-chinmina-url"
+  export CHINMINA_TOKEN_LIBRARY_FUNCTION_AUDIENCE="default"
+}
+
+teardown() {
+  unstub curl
+  rm -rf $CACHE_FILE
+
```
[the remark this answers is not in the document]
Comment (liamstevens) [reply tests/chinmina_token.bats:27]: Seems low risk due to being in a test, implement at your discretion

File: bin/chinmina_token [hunk 26]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+set -ueo pipefail
+
+# exported by plugin "environment" hook
+url="$CHINMINA_TOKEN_LIBRARY_FUNCTION_CHINMINA_URL"
+audience="$CHINMINA_TOKEN_LIBRARY_FUNCTION_AUDIENCE"
+
```
[the remark this answers is not in the document]
Comment (liamstevens) [reply bin/chinmina_token:7]: These vars are set by the plugin configuration and are both required at that level. There was some discussion around **how** that gets set for **use** since this is a library plugin, but I think this can be ignored.

File: tests/environment.bats [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+load "${BATS_PLUGIN_PATH}/load.bash"
+
+# Uncomment the following line to debug stub failures
+# export CURL_STUB_DEBUG=/dev/tty
+# export DOCKER_STUB_DEBUG=/dev/tty
+
+run_environment() {
+  echo "Running environment hook with args: $*"
+  run bash -c "source $* && env"
+}
```
[the remark this answers is not in the document]
Comment (anshbajpai01) [reply tests/environment.bats:12]: @liamstevens should this be taken into consideration ?

File: bin/chinmina_token [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+profile_name=$1
+
+# File store for the cached OIDC token
+cache_file="/tmp/oidc_auth_token_${BUILDKITE_JOB_ID}.cache"
+
+# Caches the OIDC token for 5 minutes.
+if [[ -f "${cache_file}" && -s "${cache_file}" && $(find "${cache_file}" -mmin -5) ]]; then
+  oidc_auth_token="$(< "${cache_file}")"
+else
+  # timings are output to stderr, which Git ignores.
```
[the remark this answers is not in the document]
Comment (n-tucker) [reply bin/chinmina_token:18]: Not a bad idea, however considering that these tokens only last for 5 mins and would require access to the Buildkite instance it's not critical IMO

File: .envrc [hunk 29]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+export GITHUB_TOKEN=""
+
+# Override environment variable defaults in .envrc.private
+source_env_if_exists .envrc.private
```
[the remark this answers is not in the document]
Comment (n-tucker) [reply .envrc:6]: This is intentional as sensitive secrets should be overwritten in `.envrc.private`

File: tests/environment.bats (hunk 27 above)
[the remark this answers is not in the document]
Comment (n-tucker) [reply tests/environment.bats:12]: It's very unlikely that `PWD` will ever contain spaces, if the suggested fix works then happy to include it

[the remark this answers is not in the document]
Comment (anshbajpai01) [reply tests/environment.bats:12]: @n-tucker the fix works for the current test cases, should any other scenario be included as well ?

File: tests/chinmina_token.bats (hunk 25 above)
[the remark this answers is not in the document]
Comment (anshbajpai01) [reply tests/chinmina_token.bats:27]: According to the test setup, there shouldn't be a scenario where "$CACHE_FILE" is empty 
```
  #Path to the cache file 
  CACHE_FILE="/tmp/oidc_auth_token_${BUILDKITE_JOB_ID}.cache"
  echo "$OIDC_TOKEN" > "$CACHE_FILE"
  export CACHE_FILE
```

File: README.md [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+
+1. Fork the repo
+2. Make the changes
+3. Run the tests and linter
+4. Commit and push your changes
+5. Send a pull request
+
+[chinmina-bridge]: https://chinmina.github.io/introduction/
+[chinmina-integration]: https://chinmina.github.io/guides/buildkite-integration/
+[organization-profiles]: https://chinmina.github.io/reference/organization-profile/
```
Comment (therealvio) [inline README.md:85]: Are these links supposed to be visible? In the rich view/markdown preview, they're not.

File: README.md [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+
+```yml
+steps:
+  - plugins:
+      - chinmina/chinmina-token#v1.0.0:
+          chinmina-url: "https://chinmina-bridge-url"
+          audience: "chinmina:your-github-organization"
+```
+
+To use the function to get a GitHub token, then fetch a private GitHub release
```
Comment (therealvio) [inline README.md:28]:
```suggestion
To get a GitHub token, then fetch a private GitHub release
```

File: tests/environment.bats (hunk 27 above)
[the remark this answers is not in the document]
Comment (liamstevens) [reply tests/environment.bats:12]: Let's not overcook the tests. We've got strong control over the environment in which they are executed. 

This can be closed, especially if you've implemented the suggestion and it still works.

File: README.md (hunk 30 above)
Comment (liamstevens) [reply README.md:85]: No, these are reference-style links.

See: https://www.markdownguide.org/basic-syntax/#reference-style-links

Comment (therealvio) [reply README.md:85]: Ah, I understand the implementation. They're re-usable references.

**Review by therealvio — APPROVED** [review]
